From 9e77e89441872f580eb4a69d0e8c27edd99150ee Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Tue, 22 Dec 2015 14:58:14 +0200
Subject: [PATCH 0615/1240] fix: dts: fixed base address calculation in FDT
 hierarchy

- The method calculate the base only if the node is internal-reg.
- This patch update the method to return the ranges for every node

Change-Id: I99f3bd393f7a14b7065342e31195c71b7b2e72ab
Signed-off-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/26217
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
---
 arch/arm/cpu/mvebu-common/fdt.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/arch/arm/cpu/mvebu-common/fdt.c b/arch/arm/cpu/mvebu-common/fdt.c
index 876ee87..3e64c87 100644
--- a/arch/arm/cpu/mvebu-common/fdt.c
+++ b/arch/arm/cpu/mvebu-common/fdt.c
@@ -42,17 +42,19 @@ void *fdt_get_regs_base(const void *blob, int node, uintptr_t reg)
 		return (void *)reg;
 
 	ranges = &bus_info;
+
 #if defined(CONFIG_MVEBU_SPL_DIFFRENT_BASE_ADDR) && defined(CONFIG_SPL_BUILD)
-	if (strncmp(fdt_get_name(blob, parent, NULL), "internal-regs", 13) == 0)
-		if (fdtdec_get_int_array_count(blob, parent, "ranges-spl", (u32 *)ranges, 3) == -FDT_ERR_NOTFOUND)
+	/* Part of Marvell SoC use different internal address in SPL/U-Boot */
+	if ((strncmp(fdt_get_name(blob, parent, NULL), "internal-regs", 13) == 0) &&
+		(fdtdec_get_int_array_count(blob, parent, "ranges-spl", (u32 *)ranges, 3) == -FDT_ERR_NOTFOUND))
 			return NULL;
-#else
+#endif
 	/* if there is no "ranges" property in current node then skip it */
 	if (fdtdec_get_int_array_count(blob, parent, "ranges", (u32 *)ranges, 3) == -FDT_ERR_NOTFOUND)
 		return fdt_get_regs_base(blob, parent, reg);
-#endif
 	if (reg < ranges->child_bus_address || reg > (ranges->child_bus_address + ranges->size))
 		printf("%s register base address not in the ranges\n", fdt_get_name(blob, node, NULL));
+
 	reg = reg + ranges->parent_bus_address - ranges->child_bus_address;
 	return fdt_get_regs_base(blob, parent, reg);
 }
-- 
1.9.1

